spynnaker/pyNN/models/neuron/population_machine_vertex.py

```diff
@@ -384,6 +397,227 @@ def get_binary_file_name(self):
+        "machine_graph": "MemoryMachineGraph",
+        "routing_info": "MemoryRoutingInfos",
+        "data_n_time_steps": "DataNTimeSteps",
+        "n_key_map": "MemoryMachinePartitionNKeysMap"
+    })
+    @overrides(
+        AbstractGeneratesDataSpecification.generate_data_specification,
+        additional_arguments={
+            "machine_time_step", "time_scale_factor",
+            "application_graph", "machine_graph", "routing_info",
+            "data_n_time_steps", "n_key_map"
+        })
+    def generate_data_specification(
+            self, spec, placement, machine_time_step, time_scale_factor,
+            application_graph, machine_graph, routing_info, data_n_time_steps,
+            n_key_map):
+        """
+        :param machine_time_step: (injected)
+        :param time_scale_factor: (injected)
+        :param application_graph: (injected)
+        :param machine_graph: (injected)
+        :param routing_info: (injected)
+        :param data_n_time_steps: (injected)
+        :param n_key_map: (injected)
+        """
+        # pylint: disable=too-many-arguments, arguments-differ
+
+        spec.comment("\n*** Spec for block of {} neurons ***\n".format(
+            self._app_vertex.neuron_impl.model_name))
+
+        # Reserve memory regions
+        self._reserve_memory_regions(spec, machine_graph, n_key_map)
+
+        # Declare random number generators and distributions:
+        # TODO add random distribution stuff
+        # self.write_random_distribution_declarations(spec)
+
+        # Get the key
+        key = routing_info.get_first_key_from_pre_vertex(
+            self, constants.SPIKE_PARTITION_ID)
+
+        # Write the setup region
+        spec.switch_write_focus(POPULATION_BASED_REGIONS.SYSTEM.value)
+        spec.write_array(simulation_utilities.get_simulation_header_array(
+            self.__binary_file_name, machine_time_step, time_scale_factor))
+
+        # Write the neuron recording region
+        self._app_vertex.neuron_recorder.write_neuron_recording_region(
+            spec, POPULATION_BASED_REGIONS.NEURON_RECORDING.value,
+            self.vertex_slice, data_n_time_steps)
+
+        # Write the neuron parameters
+        self._write_neuron_parameters(
+            spec, key, constants.POPULATION_BASED_REGIONS.NEURON_PARAMS.value)
+
+        # write profile data
+        profile_utils.write_profile_region_data(
+            spec, POPULATION_BASED_REGIONS.PROFILING.value,
+            self._app_vertex.n_profile_samples)
+
+        # Get the weight_scale value from the appropriate location
+        weight_scale = self._app_vertex.neuron_impl.get_global_weight_scale()
+
+        # allow the synaptic matrix to write its data spec-able data
+        self._app_vertex.synapse_manager.write_data_spec(
+            spec, self._app_vertex, self.vertex_slice, self, machine_graph,
+            application_graph, routing_info, weight_scale, machine_time_step)
```

Comment:
Application_graph usage needs to come out here. But out of scope for this PR I think.

Reply:
i agree, and i agree its out of scope of pr. thats a seperate problem in of itself

Reply:
https://github.com/SpiNNakerManchester/sPyNNaker/issues/926